pkg/zabbix_api.go

```diff
@@ -275,3 +280,336 @@ func isNotAuthorized(message string) bool {
+
+	return ds.ZabbixRequest(ctx, dsInfo, "host.get", params)
+}
+
+func (ds *ZabbixDatasource) getAllGroups(ctx context.Context, dsInfo *datasource.DatasourceInfo) (*simplejson.Json, error) {
+	params := zabbixParams{Output: &zabbixParamOutput{Fields: []string{"name"}}, SortField: "name", RealHosts: true}
+
+	return ds.ZabbixRequest(ctx, dsInfo, "hostgroup.get", params)
+}
+
+func (ds *ZabbixDatasource) queryNumericDataForItems(ctx context.Context, tsdbReq *datasource.DatasourceRequest, items zabbix.Items, jsonQueries []*simplejson.Json, useTrend bool) ([]*datasource.TimeSeries, error) {
+	valueType := ds.getTrendValueType(jsonQueries)
+	consolidateBy := ds.getConsolidateBy(jsonQueries)
+
+	if consolidateBy == "" {
+		consolidateBy = valueType
+	}
+
+	history, err := ds.getHistotyOrTrend(ctx, tsdbReq, items, useTrend)
+	if err != nil {
+		return nil, err
+	}
+
+	return convertHistory(history, items)
+}
+
+func (ds *ZabbixDatasource) getTrendValueType(jsonQueries []*simplejson.Json) string {
+	var trendFunctions []string
+	var trendValueFunc string
+
+	// TODO: loop over populated categories
+	for _, j := range new(categories).Trends {
+		trendFunctions = append(trendFunctions, j["name"].(string))
+	}
+	for _, k := range jsonQueries[0].Get("functions").MustArray() {
+		for _, j := range trendFunctions {
+			if j == k.(map[string]interface{})["def"].(map[string]interface{})["name"] {
+				trendValueFunc = j
+			}
+		}
+	}
+
+	if trendValueFunc == "" {
+		trendValueFunc = "avg"
+	}
+
+	return trendValueFunc
+}
+
+func (ds *ZabbixDatasource) getConsolidateBy(jsonQueries []*simplejson.Json) string {
+	var consolidateBy string
+
+	for _, k := range jsonQueries[0].Get("functions").MustArray() {
+		if k.(map[string]interface{})["def"].(map[string]interface{})["name"] == "consolidateBy" {
+			defParams := k.(map[string]interface{})["def"].(map[string]interface{})["params"].([]interface{})
+			if len(defParams) > 0 {
+				consolidateBy = defParams[0].(string)
+			}
+		}
+	}
+	return consolidateBy
+}
+
+func (ds *ZabbixDatasource) getHistotyOrTrend(ctx context.Context, tsdbReq *datasource.DatasourceRequest, items zabbix.Items, useTrend bool) (zabbix.History, error) {
+	allHistory := zabbix.History{}
+
+	timeRange := tsdbReq.GetTimeRange()
+	groupedItems := map[int]zabbix.Items{}
+
+	for _, j := range items {
+		groupedItems[j.ValueType] = append(groupedItems[j.ValueType], j)
+	}
+
+	for k, l := range groupedItems {
+		var itemids []string
+		for _, m := range l {
+			itemids = append(itemids, m.ID)
+		}
+
+		params := zabbixParams{
+			Output:    &zabbixParamOutput{Mode: "extend"},
+			SortField: "clock",
+			SortOrder: "ASC",
+			ItemIDs:   itemids,
+			TimeFrom:  timeRange.GetFromEpochMs() / 1000,
+			TimeTill:  timeRange.GetToEpochMs() / 1000,
+		}
+
+		var response *simplejson.Json
+		var err error
+		if useTrend {
+			response, err = ds.ZabbixRequest(ctx, tsdbReq.GetDatasource(), "trend.get", params)
+		} else {
+			params.History = k
```

Comment:
History should be a pointer in order to prevent it to be omitted.
```suggestion
			params.History = &k
```

Reply:
9511524